lib/OrdinaryDiffEqRosenbrock/src/rosenbrock_perform_step.jl

```diff
@@ -1246,138 +1228,89 @@ end
-    du = f(u, p, t + c2 * dt)
-    OrdinaryDiffEqCore.increment_nf!(integrator.stats, 1)
-
-    if mass_matrix === I
-        linsolve_tmp = du + dtd2 * dT + dtC21 * k1
-    else
-        linsolve_tmp = du + dtd2 * dT + mass_matrix * (dtC21 * k1)
-    end
-
-    k2 = _reshape(W \ -_vec(linsolve_tmp), axes(uprev))
-    integrator.stats.nsolve += 1
-    u = uprev + a31 * k1 + a32 * k2
-    du = f(u, p, t + c3 * dt)
-    OrdinaryDiffEqCore.increment_nf!(integrator.stats, 1)
-
-    if mass_matrix === I
-        linsolve_tmp = du + dtd3 * dT + (dtC31 * k1 + dtC32 * k2)
-    else
-        linsolve_tmp = du + dtd3 * dT + mass_matrix * (dtC31 * k1 + dtC32 * k2)
-    end
-
-    k3 = _reshape(W \ -_vec(linsolve_tmp), axes(uprev))
-    integrator.stats.nsolve += 1
-    u = uprev + a41 * k1 + a42 * k2 + a43 * k3
-    du = f(u, p, t + c4 * dt)
-    OrdinaryDiffEqCore.increment_nf!(integrator.stats, 1)
+    # Initialize k arrays
+    num_stages = size(A,1)
 
-    if mass_matrix === I
-        linsolve_tmp = du + dtd4 * dT + (dtC41 * k1 + dtC42 * k2 + dtC43 * k3)
-    else
-        linsolve_tmp = du + dtd4 * dT + mass_matrix * (dtC41 * k1 + dtC42 * k2 + dtC43 * k3)
-    end
-
-    k4 = _reshape(W \ -_vec(linsolve_tmp), axes(uprev))
-    integrator.stats.nsolve += 1
-    u = uprev + a51 * k1 + a52 * k2 + a53 * k3 + a54 * k4
-    du = f(u, p, t + dt)
-    OrdinaryDiffEqCore.increment_nf!(integrator.stats, 1)
+    # Loop for stages
+    for stage in 1:num_stages
+        u = uprev
+        for i in 1:stage-1
+            u = @.. u + A[stage, i] * ks[i]
+        end
 
-    if mass_matrix === I
-        linsolve_tmp = du + (dtC52 * k2 + dtC54 * k4 + dtC51 * k1 + dtC53 * k3)
-    else
-        linsolve_tmp = du +
-                       mass_matrix * (dtC52 * k2 + dtC54 * k4 + dtC51 * k1 + dtC53 * k3)
-    end
+        du = f(u, p, t + c[stage] * dt)
+        OrdinaryDiffEqCore.increment_nf!(integrator.stats, 1)
 
-    k5 = _reshape(W \ -_vec(linsolve_tmp), axes(uprev))
-    integrator.stats.nsolve += 1
-    u = u + k5
-    du = f(u, p, t + dt)
-    OrdinaryDiffEqCore.increment_nf!(integrator.stats, 1)
+        # Compute linsolve_tmp for current stage
+        linsolve_tmp1 = zero(du)
+        if mass_matrix === I
+            for i in 1:stage-1
+                linsolve_tmp1 = @.. linsolve_tmp1 + dtC[stage, i] * ks[i]
+            end
+        else
+            for i in 1:stage-1
+                linsolve_tmp1 = @.. linsolve_tmp1 + dtC[stage, i] * ks[i]
+            end
+            linsolve_tmp1 = mass_matrix * linsolve_tmp1
+        end
+        linsolve_tmp = @.. du + dtd[stage] * dT + linsolve_tmp1
 
-    if mass_matrix === I
-        linsolve_tmp = du + (dtC61 * k1 + dtC62 * k2 + dtC65 * k5 + dtC64 * k4 + dtC63 * k3)
-    else
-        linsolve_tmp = du +
-                       mass_matrix *
-                       (dtC61 * k1 + dtC62 * k2 + dtC65 * k5 + dtC64 * k4 + dtC63 * k3)
+        ks[stage] = _reshape(W \ -_vec(linsolve_tmp), axes(uprev))
```

Comment:
mutation of this form is going to break downstream autodiff with some algorithms.

Reply:
It needs to be something like a map.

Reply:
this can't be something like a map (since each `k` needs the previous `ks`). Would using a Tuple with Accessors.jl work?

Reply:
Yes

Reply:
fixed.